File: packages/capacitor-mesh/ios/MeshWireProtocol.swift
import Foundation
import CryptoKit

enum MeshWireMode: Equatable {
    case unknown
    case legacy
    case prm1
}

enum MeshWireNegotiation {
    static let magic = Data([0x50, 0x52, 0x4d, 0x31])
    static let capabilityData = Data([0x50, 0x52, 0x4d, 0x31, 0x01])

    static func detectMode(_ data: Data) -> MeshWireMode {
        if data.count >= magic.count && data.prefix(magic.count) == magic {
            return .prm1
        }
        guard let first = data.first else { return .unknown }
        if first == 0x7b || first == 0x5b { // `{` or `[`: legacy JSON
            return .legacy
        }
        return .unknown
    }
}

/// Versioned stream framing shared with the Android mesh plugin.
/// CoreBluetooth writes may split this stream at arbitrary boundaries; only a
/// complete, checksum-verified logical JSON message is emitted to JavaScript.
enum MeshWireProtocol {
    static let magic = Data([0x50, 0x52, 0x4d, 0x31]) // "PRM1"
    static let version: UInt8 = 1
    static let headerBytes = 32
    static let maxMessageBytes = 1024 * 1024
    private static let tagBytes = 16

    enum ProtocolError: Error {
        case messageTooLarge
        case unsupportedVersion
        case integrityFailure
    }

    static func encode(_ payload: Data) throws -> Data {
        guard payload.count <= maxMessageBytes else { throw ProtocolError.messageTooLarge }
        var header = Data()
        header.append(magic)
        header.append(version)
        header.append(0) // flags
        header.append(contentsOf: [0, 0]) // reserved
        appendUInt32(&header, UInt32(payload.count))
        appendUInt32(&header, crc32(payload))
        header.append(Data(SHA256.hash(data: payload).prefix(tagBytes)))
        return header + payload
    }

    final class Decoder {
        private var buffer = Data()

        func reset() {
            buffer.removeAll(keepingCapacity: false)
        }

        func append(_ chunk: Data) throws -> [Data] {
            buffer.append(chunk)
            var messages: [Data] = []
            while buffer.count >= headerBytes {
                alignToMagic()
                if buffer.count < headerBytes { break }
                guard buffer[4] == version else {
                    reset()
                    throw ProtocolError.unsupportedVersion
                }
                let payloadLength = Int(readUInt32(buffer, offset: 8))
                guard payloadLength <= maxMessageBytes else {
                    reset()
                    throw ProtocolError.messageTooLarge
                }
                let total = headerBytes + payloadLength
                if buffer.count < total { break }
                let payload = buffer.subdata(in: headerBytes..<total)
                let expectedChecksum = readUInt32(buffer, offset: 12)
                let expectedTag = buffer.subdata(in: 16..<(16 + tagBytes))
                buffer.removeSubrange(0..<total)
                let actualTag = Data(SHA256.hash(data: payload).prefix(tagBytes))
                guard expectedChecksum == crc32(payload), expectedTag == actualTag else {
                    reset()
                    throw ProtocolError.integrityFailure
                }
                messages.append(payload)
            }
            return messages
        }

        private func alignToMagic() {
            if buffer.starts(with: MeshWireProtocol.magic) { return }
            if let range = buffer.range(of: MeshWireProtocol.magic, options: [], in: buffer.startIndex..<buffer.endIndex), range.lowerBound > 0 {
                buffer.removeSubrange(0..<range.lowerBound)
            } else if buffer.range(of: MeshWireProtocol.magic) == nil {
                buffer = Data(buffer.suffix(MeshWireProtocol.magic.count - 1))
            }
        }
    }

    private static func appendUInt32(_ data: inout Data, _ value: UInt32) {
        data.append(UInt8((value >> 24) & 0xff))
        data.append(UInt8((value >> 16) & 0xff))
        data.append(UInt8((value >> 8) & 0xff))
        data.append(UInt8(value & 0xff))
    }

    private static func readUInt32(_ data: Data, offset: Int) -> UInt32 {
        (UInt32(data[offset]) << 24)
            | (UInt32(data[offset + 1]) << 16)
            | (UInt32(data[offset + 2]) << 8)
            | UInt32(data[offset + 3])
    }

    private static func crc32(_ data: Data) -> UInt32 {
        var crc: UInt32 = 0xffffffff
        for byte in data {
            crc ^= UInt32(byte)
            for _ in 0..<8 {
                crc = (crc >> 1) ^ ((crc & 1) == 1 ? 0xedb88320 : 0)
            }
        }
        return crc ^ 0xffffffff
    }
}

File: packages/capacitor-mesh/android/src/main/java/com/praeventio/mesh/MeshWireProtocol.kt
package com.praeventio.mesh

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest
import java.util.zip.CRC32

/**
 * Versioned stream framing shared by the Android/iOS mesh plugins.
 *
 * A logical JSON packet is encoded once, then the resulting bytes may be split
 * at arbitrary GATT write boundaries. The receiver emits only after it has the
 * complete declared message and the CRC/SHA tag both match. GATT writes are
 * deliberately not treated as packet boundaries.
 */
internal object MeshWireProtocol {
    private val MAGIC = byteArrayOf(0x50, 0x52, 0x4d, 0x31) // "PRM1"
    const val VERSION = 1
    const val HEADER_BYTES = 32
    const val MAX_MESSAGE_BYTES = 1024 * 1024
    private const val TAG_BYTES = 16

    class ProtocolException(message: String) : IllegalArgumentException(message)

    fun encode(payload: ByteArray): ByteArray {
        requireSize(payload.size)
        val header = ByteArray(HEADER_BYTES)
        MAGIC.copyInto(header, 0)
        header[4] = VERSION.toByte()
        // header[5..7] flags/reserved remain zero.
        putUInt32(header, 8, payload.size.toLong())
        putUInt32(header, 12, crc32(payload))
        sha256(payload).copyInto(header, 16, 0, TAG_BYTES)
        return concat(header, payload)
    }

    class Decoder {
        private var buffer = ByteArray(0)

        fun reset() {
            buffer = ByteArray(0)
        }

        fun append(chunk: ByteArray): List<ByteArray> {
            if (chunk.isNotEmpty()) buffer = concat(buffer, chunk)
            val messages = ArrayList<ByteArray>()
            while (buffer.size >= HEADER_BYTES) {
                alignToMagic()
                if (buffer.size < HEADER_BYTES) break
                if (buffer[4].toInt() and 0xff != VERSION) {
                    reset()
                    throw ProtocolException("unsupported mesh wire version")
                }
                val payloadLength = readUInt32(buffer, 8)
                if (payloadLength > MAX_MESSAGE_BYTES) {
                    reset()
                    throw ProtocolException("mesh message exceeds maximum")
                }
                val total = HEADER_BYTES + payloadLength.toInt()
                if (buffer.size < total) break
                val payload = buffer.copyOfRange(HEADER_BYTES, total)
                val expectedChecksum = readUInt32(buffer, 12)
                val expectedTag = buffer.copyOfRange(16, 16 + TAG_BYTES)
                buffer = buffer.copyOfRange(total, buffer.size)
                if (expectedChecksum != crc32(payload)
                    || !expectedTag.contentEquals(sha256(payload).copyOf(TAG_BYTES))) {
                    reset()
                    throw ProtocolException("mesh message integrity check failed")
                }
                messages.add(payload)
            }
            return messages
        }

        private fun alignToMagic() {
            if (hasMagic(buffer, 0)) return
            val index = findMagic(buffer)
            buffer = if (index >= 0) {
                buffer.copyOfRange(index, buffer.size)
            } else {
                buffer.copyOfRange(
                    maxOf(0, buffer.size - (MAGIC.size - 1)),
                    buffer.size,
                )
            }
        }
    }

    private fun requireSize(size: Int) {
        if (size > MAX_MESSAGE_BYTES) {
            throw ProtocolException("mesh message exceeds maximum")
        }
    }

    private fun hasMagic(input: ByteArray, offset: Int): Boolean {
        if (offset < 0 || offset + MAGIC.size > input.size) return false
        for (i in MAGIC.indices) if (input[offset + i] != MAGIC[i]) return false
        return true
    }

    private fun findMagic(input: ByteArray): Int {
        for (i in 1..(input.size - MAGIC.size)) if (hasMagic(input, i)) return i
        return -1
    }

    private fun readUInt32(input: ByteArray, offset: Int): Long {
        return ((input[offset].toLong() and 0xff) shl 24) or
            ((input[offset + 1].toLong() and 0xff) shl 16) or
            ((input[offset + 2].toLong() and 0xff) shl 8) or
            (input[offset + 3].toLong() and 0xff)
    }

    private fun putUInt32(output: ByteArray, offset: Int, value: Long) {
        ByteBuffer.wrap(output, offset, 4)
            .order(ByteOrder.BIG_ENDIAN)
            .putInt(value.toInt())
    }

    private fun sha256(input: ByteArray): ByteArray =
        MessageDigest.getInstance("SHA-256").digest(input)

    private fun crc32(input: ByteArray): Long {
        val crc = CRC32()
        crc.update(input)
        return crc.value
    }

    private fun concat(first: ByteArray, second: ByteArray): ByteArray {
        val output = ByteArray(first.size + second.size)
        first.copyInto(output, 0)
        second.copyInto(output, first.size)
        return output
    }
}
